### ISSUE TITLE
False positive in `one-var` for successive `for-in` and `for-of` loops

### ISSUE THREAD

**Issue by impinball** [eslint/eslint#3073]
Minimal repro case:

```sh
# Command line:
eslint --no-eslintrc --rule '{"one-var":[2,{"uninitialized": "always"}]}' file.js
```

```js
// file.js
for (var i in foo) {}
for (var j in foo) {}
```

Output:

```
tmp.js
  2:6  error  Combine this with the previous 'var' statement with uninitialized variables  one-var

✖ 1 problem (1 error, 0 warnings)
```

Same problem with with `let` and `const`, and with `for-of` loops (both with `--env es6`).

```js
// file.js
for (let i of foo) {}
for (let j of foo) {}
```

As for `let`/`const`, the semantics are different, so it should never trigger the warning, but it probably shouldn't for `var`, either.

**Comment by btmills** [comment]
In the case of `var`, the preferred style would be to declare both variables at the top of the scope instead of inside the loop. For `let`, however, that looks correct since the loop would create a new scope. I'd say that part is a bug with how the rule determines scopes.

**Comment by mysticatea** [comment]
I will try to fix this.

**Comment by impinball** [comment]
Does that also fix the `for ... in` case?

**Comment by impinball** [comment]
And in terms of how variables are defined, the semantics are identical between the two, IIRC.

**Comment by jrvidal** [comment]
The `for-in` case is correctly reported. The difference is not the loop type itself, but the fact that you are using `var` instead of `let`.

**Comment by mysticatea** [comment]
@impinball 

```js
for (var a in []) { }
for (let b in []) { }
```

This `a` is a variable in the global scope.
But this `b` is a variable in the scope of `for-in` loop.

So the two are different in semantics.
This time, the logic which makes scopes of `for-in`/`for-of` for `let`/`const` was wrong.

----

I have one question.

```js
for (var a in b) { }
```

This `a` is initialized or uninitialized?

Currently, it's handled uninitialized. But it looks initialized to me.

**Comment by mysticatea** [comment]
Ah, I'm sorry, I misstook understanding `between the two`.

**Comment by impinball** [comment]
According to the spec, as far back as ES3, it is initialized each iteration to the current priory (although successive iterations are visibly equivalent to the user as assignment). See section 12.6.4 in the ES3/5/5.1 specs. ES6 has similar semantics, but it's not as explicit within the spec itself because has to handle the destructuring special case (section 13.7.5.13).

**Comment by impinball** [comment]
It's actually being initialized at declaration time each iteration, though.

**Comment by impinball** [comment]
So could this be reopened?
